publisher/release/marker_aws.go

```diff
@@ -0,0 +1,204 @@
+	}
+
+	creds := stscreds.NewCredentials(sess, s3Config.RoleARN, func(p *stscreds.AssumeRoleProvider) {})
+	awsCfg := aws.Config{
+		Credentials: creds,
+		Region:      aws.String(s3Config.Region),
+	}
+
+	return &markerAWS{
+		client:       s3.New(sess, &awsCfg),
+		conf:         s3Config,
+		timeProvider: RealTimeProvider{},
+		logfn:        logfn,
+	}, nil
+}
+
+// Start will:
+// load all the markers from the file
+// append a new started marker
+// write the markers back to the file
+func (s *markerAWS) Start(appName string, tag string, runID string, repoName string, schema string, schemaURL string) (Mark, error) {
+	s.logfn("[marker] starting %s", appName)
+	markers, err := s.readMarkers()
+	if err != nil {
+		if !isNoSuchKeyError(err) {
+			return Mark{}, err
+		}
+	}
+
+	mark := Mark{
+		AppName:   appName,
+		Tag:       tag,
+		RunID:     runID,
+		RepoName:  repoName,
+		Schema:    schema,
+		SchemaURl: schemaURL,
+		Start:     CustomTime{s.now()},
+	}
+
+	markers = append(markers, mark)
+	if err = s.writeMarkers(markers); err != nil {
+		return mark, err
+	}
+
+	return mark, nil
+}
+
+// End will:
+// load all the markers from the file
+// find the last started marker
+// append the end time to the last started marker
+// write the markers back to the file
+func (s *markerAWS) End(mark Mark) error {
+	s.logfn("[marker] ending %s", mark.AppName)
+	if mark.Start.IsZero() {
+		return ErrNotStartedMark
+	}
+
+	markers, err := s.readMarkers()
+	if err != nil {
+		return err
+	}
+	//ensure the latest marker is the one being ended
+	if len(markers) == 0 {
+		return ErrNoStartedMarkersFound
+	}
+
+	lastMarker := markers[len(markers)-1]
+	if !lastMarker.End.IsZero() {
+		return ErrLastMarkerEnded
+	}
+
+	if lastMarker.AppName != mark.AppName || !lastMarker.Start.Equals(mark.Start) {
+		return fmt.Errorf("%w started:%s appName:%s", ErrNoStartedMarkerFoundForApp, mark.Start, mark.AppName)
+	}
+
+	lastMarker.End = CustomTime{s.now()}
+	markers[len(markers)-1] = lastMarker
+
+	err = s.writeMarkers(markers)
+	if err != nil {
+		return fmt.Errorf("%w: %w", ErrCannotWriteMarkerFile, err)
+	}
+
+	return nil
+}
+
+func (s *markerAWS) writeMarkers(markers []Mark) error {
+	markersBytes, err := json.MarshalIndent(markers, "", "  ")
+	if err != nil {
+		return fmt.Errorf("cannot encode marker file: %w", err)
+	}
+
+	s.logfn("[marker] writing bucket:%s key:%s", s.conf.Bucket, s.markerPath())
+	_, err = s.client.PutObject(&s3.PutObjectInput{
+		Bucket: aws.String(s.conf.Bucket),
+		Key:    aws.String(s.markerPath()),
+		Body:   aws.ReadSeekCloser(bytes.NewReader(markersBytes)),
+	})
+	if err != nil {
+		return fmt.Errorf("cannot write marker file: %w", err)
+	}
+
+	return nil
+}
+
+func (s *markerAWS) readMarkers() ([]Mark, error) {
+	objOutput, err := s.client.GetObject(&s3.GetObjectInput{
+		Bucket: aws.String(s.conf.Bucket),
+		Key:    aws.String(s.markerPath()),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("cannot read marker file: %w", err)
+	}
```

Comment:
Is the markers supposed to exist or will the action create it the first time?

Reply:
It will be created if it doesn't exist https://github.com/newrelic/infrastructure-publish-action/pull/170/files#diff-746e45e2647cf17fb02b10957238994fabb26c6025dcb63eec2a0e14e7a55094R89

Reply:
I totally missed that!

```go
	if err != nil {
		if !isNoSuchKeyError(err) {
			return Mark{}, err
		}
```

Shall we add a comment there to make more obvious?

Reply:
👍  Added in 27048d7f762efc973a945519ab82a79e96621c3a